src/Raven.Server/Documents/Handlers/Processors/Batches/BatchHandlerProcessorForBulkDocs.cs

```diff
@@ -103,8 +104,10 @@ public static async Task WaitForIndexesAsync(DocumentDatabase database, TimeSpan
                         continue;
 
                     hadStaleIndexes = true;
-
-                    await waitForIndexItem.WaitForIndexing.WaitForIndexingAsync(waitForIndexItem.IndexBatchAwaiter);
+                    using (waitForIndexItem.IndexBatchAwaiter.Register(token))
+                    {
+                        await waitForIndexItem.WaitForIndexing.WaitForIndexingAsync(waitForIndexItem.IndexBatchAwaiter);
+                    }
```

Comment:
instead let's pass the token to `WaitForIndexingAsync` and do something like this in the `FrozenAwaiter` :

```
 public async Task<bool> WaitAsync(CancellationToken token)
 {
     var tcs = _tcs;
     await using (token.Register(() => tcs.TrySetCanceled()))
     {
         return await _tcs.Task.ConfigureAwait(false);
     }
 }
```

Reply:
or maybe even better
```
return _parent.WaitAsync(token);
```

Reply:
Done